implementations/elixir/ockam/ockam_hub/lib/hub/service/token_lease_manager.ex

```diff
@@ -1,130 +1,242 @@
-    set_expiration(self(), token_id, expiration)
-    {:noreply, state}
+  def handle_message({:set_ttl, token_id, ttl}, state) do
+    set_ttl(self(), token_id, ttl)
+    {:ok, state}
   end
 
-  def handle_info(:set_all_expiration, state) do
+  def handle_message(:set_all_ttl, %{storage_service: storage_service} = state) do
+    Logger.info("setting all ttls")
     lease_manager_pid = self()
 
     Task.start(fn ->
-      Enum.map(get_all_leases(), fn lease ->
-        set_expiration(lease_manager_pid, lease[:token_id], lease[:expiration])
+      {:ok, leases} = storage_service.get_all()
+
+      Enum.map(leases, fn lease ->
+        ttl = new_ttl(lease.ttl, lease.issued)
+        set_ttl(lease_manager_pid, lease.id, ttl)
       end)
     end)
 
-    {:noreply, state}
+    {:ok, state}
   end
 
-  def handle_info(_message, state) do
-    {:noreply, state}
+  def handle_message({:auto_revoke, token_id}, state) do
+    Logger.info("revoking #{token_id} token")
+    Task.start(fn -> do_revoke(state[:cloud_service], state[:storage_service], token_id) end)
+    {:ok, state}
   end
 
-  defp get_cloud_service(name) do
-    case name do
-      :influxdb -> Ockam.TokenLeaseManager.CloudServiceInfluxdb
-      _other -> Ockam.TokenLeaseManager.CloudServiceInfluxdb
-    end
+  defp process_request(received_payload, state) do
+    {_prefix, payload} = get_payload_prefix(received_payload)
+
+    response = process_encoded_payload(payload, state)
+
+    length =
+      case String.length(response) do
+        l when l > 128 and l < 255 ->
+          :binary.encode_unsigned(l) <> <<1>>
+
+        l ->
+          :binary.encode_unsigned(l)
+      end
+
+    length <> response
   end
 
-  defp get_storage_service(name) do
-    # TODO: storage system
-    case name do
-      _other -> :storage
-    end
+  defp get_payload_prefix(payload), do: get_payload_prefix(payload, <<>>)
+
+  defp get_payload_prefix("{" <> _rest = payload, current_prefix) do
+    {current_prefix, payload}
   end
 
-  defp set_expiration(lease_manager_pid, token_id, expiration) do
-    Process.send_after(lease_manager_pid, {:revoke, token_id}, expiration)
+  defp get_payload_prefix(<<new::size(8)>> <> payload, current_prefix) do
+    get_payload_prefix(payload, current_prefix <> <<new>>)
   end
 
-  defp save_lease(_lease) do
-    # TODO
+  defp get_cloud_service(name) do
+    with mod <- get_cloud_service_module(name),
+         {:ok, _pid} <- mod.start_link() do
+      {:ok, mod}
+    else
+      error -> error
+    end
   end
 
-  defp remove_lease(_tokend_id) do
-    # TODO
+  defp get_storage_service(name, token_cloud_service, token_cloud_service_address) do
+    with mod <- get_storage_service_module(name),
+         {:ok, _pid} <- mod.start_link({token_cloud_service, token_cloud_service_address}) do
+      {:ok, mod}
+    else
+      error -> error
+    end
   end
 
-  defp get_all_leases() do
-    # TODO
-    []
+  defp get_cloud_service_module("INFLUXDB"),
```

Comment:
Since we have the pluggable behaviour, I think having a strict set of possible modules is not the best idea. 
Can we move the string to module name association out of this module (maybe to the service provider) or just use module names and not strings?

Reply:
sure